index.js

```diff
@@ -1,21 +1,72 @@
 const http = require('http')
 const rangeParser = require('range-parser')
 const mime = require('mime-types')
+const ReadyResource = require('ready-resource')
+const z32 = require('z32')
+const safetyCatch = require('safety-catch')
 
-module.exports = async function serve (drives, opts = {}) {
-  if (!(drives instanceof Map)) {
-    const drive = drives
-    drives = new Map()
-    drives.set(null, drive)
+module.exports = class ServeDrive extends ReadyResource {
+  constructor (opts = {}) {
+    super()
+
+    this.drives = new Map()
+
+    this.port = typeof opts.port !== 'undefined' ? Number(opts.port) : 7000
+    this.host = typeof opts.host !== 'undefined' ? opts.host : null
+    this.anyPort = opts.anyPort !== false
+
+    this.server = opts.server || http.createServer()
+    this.server.on('request', this._onrequest.bind(this))
   }
 
-  const port = typeof opts.port !== 'undefined' ? Number(opts.port) : 7000
-  const host = typeof opts.host !== 'undefined' ? opts.host : null
-  const anyPort = opts.anyPort !== false
+  async _open () {
+    await Promise.resolve() // Wait a tick, so you don't rely on server.address() being sync sometimes
+
+    try {
+      await listen(this.server, this.port, this.host)
+    } catch (err) {
+      if (!this.anyPort) throw err
+      if (err.code !== 'EADDRINUSE') throw err
+      await listen(this.server, 0, this.host)
+    }
+  }
 
-  const server = opts.server || http.createServer()
+  async _close () {
+    if (this.closed) return
```

Comment:
No need for this and next line if using ready-resource

Reply:
Neither for the check on this.opened I think, but not 100% sure there

Reply:
It's not needed, it was inherited from the previous code that was doing its own opening/closing